rules/handler.go

```diff
@@ -0,0 +1,391 @@
+	return HandlerOptions{propagationDelay: propDelay}
+}
+
+// MappingRuleConfig is the parts of a mapping rule that an end user would care about.
+// This struct is provided to create or make updates to the mapping rule.
+type MappingRuleConfig struct {
+	Name     string            `json:"name" validate:"nonzero"`
+	Filters  map[string]string `json:"filters" validate:"nonzero"`
+	Policies []policy.Policy   `json:"policies" validate:"nonzero"`
+}
+
+// MappingRuleUpdate contains a MappingRuleConfig along with an ID for a mapping rule.
+// The rule with that ID is meant to updated to the given config.
+type MappingRuleUpdate struct {
+	Config MappingRuleConfig `json:"config" validate:"nonzero"`
+	Name   string            `json:"name" validate:"nonzero"`
+}
+
+// RollupRuleConfig is the parts of a rollup rule that an end user would care about.
+// This struct is provided to create or make updates to the rollup rule.
+type RollupRuleConfig struct {
+	Name    string            `json:"name" validate:"nonzero"`
+	Filters map[string]string `json:"filters" validate:"nonzero"`
+	Targets []RollupTarget    `json:"targets" validate:"nonzero"`
+}
+
+// RollupRuleUpdate is a RollupRuleConfig along with an ID for a rollup rule.
+// The rule with that ID is meant to updated with a given config.
+type RollupRuleUpdate struct {
+	Config RollupRuleConfig `json:"config" validate:"nonzero"`
+	Name   string           `json:"name" validate:"nonzero"`
+}
+
+// RuleSetHandler exposes all of the methods to work with a RuleSet
+type RuleSetHandler interface {
+	// AppendMappingRule creates a new mapping rule and adds it to this ruleset.
+	AddMappingRule(MappingRuleConfig) error
+
+	// UpdateMappingRule creates a new mapping rule and adds it to this ruleset.
+	UpdateMappingRule(MappingRuleUpdate) error
+
+	// DeleteMappingRule deletes a mapping rule
+	DeleteMappingRule(ruleName string) error
+
+	// AppendRollupRule creates a new mapping rule and adds it to this ruleset.
+	AddRollupRule(RollupRuleConfig) error
+
+	// UpdateRollupRule creates a new mapping rule and adds it to this ruleset.
+	UpdateRollupRule(RollupRuleUpdate) error
+
+	// DeleteRollupRule deletes a rollup rule
+	DeleteRollupRule(ruleName string) error
+
+	// Tombstone tombstones this ruleset and all of its rules.
+	Delete() error
+
+	// Revive removes the tombstone from this ruleset. It does not revive any rules.
+	Revive() error
+
+	// Validate checks that this ruleset is good to use. e.g. not tombstoned
+	Validate() error
+
+	// Ruleset returns the latest state of the ruleset in this handler.
+	RuleSet() RuleSet
+}
+
+// ruleSetHandler is a struct that all functionality should be defined on.
+type ruleSetHandler struct {
+	opts HandlerOptions
+	rs   *ruleSet
+}
+
+func (h ruleSetHandler) RuleSet() RuleSet {
+	return h.rs
+}
+
+func (h ruleSetHandler) AddMappingRule(mrc MappingRuleConfig) error {
+	rs := h.rs
+	m, err := rs.getMappingRuleByName(mrc.Name)
+	updateTime := time.Now().UnixNano()
```

Comment:
Can we not determine the cutover time in the `updateTimestamps` function so we don't need to have these two lines in every function?